grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcMessageMarshaller.java

```diff
@@ -203,8 +206,17 @@ private <T> ByteBuf serializeProto(PrototypeMarshaller<T> marshaller, Message me
             final ByteBuf buf = alloc.buffer(serializedSize);
             boolean success = false;
             try {
-                message.writeTo(CodedOutputStream.newInstance(buf.nioBuffer(0, serializedSize)));
-                buf.writerIndex(serializedSize);
+                if (useMethodMarshaller) {
```

Comment:
I found out we also need to fix `deserializeProto`.
```java
if (useMethodMarshaller) {
    msg = (Message) marshaller.parse(new ByteBufInputStream(buf));
} else {
    stream = CodedInputStream.newInstance(buf.nioBuffer());
    msg = prototype.getParserForType().parseFrom(stream);
    ...
}
```
`unsafeWrapDeserializedBuffer` and `useMethodMarshaller` are mutually exclusive. Should we raise an `IllegalStateException` if both are enabled in the client and service builder classes?

Reply:
Okay, I'll add missing part for `deserializeProto` and check logic mutually exclusive check for both flags.

Reply:
I have added mutual exclusive check.
For fixing `deserializeProto`, Is it enough to modify code like below and testcase accordingly.

```java
private <T> Message deserializeProto(PrototypeMarshaller<T> marshaller, ByteBuf buf) throws IOException {
 //...
  final Message msg;
  if (useMethodMarshaller) {
      msg = (Message) marshaller.parse(new ByteBufInputStream(buf));
  } else {
      msg = prototype.getParserForType().parseFrom(stream);
  }
```

Reply:
@ikhoon, 
I need some help.
I've added `spiedMarshallerParseCallCnt` to check `parse` gets called when I set `useMethodMarshaller` to `true`. and then I ran test but I got NPE because I didn't return proper value for `getMessagePrototype()`. I've no idea what to return. 



```java
private static class CustomMarshallerInterceptor implements ClientInterceptor {
        private int spiedMarshallerStreamCallCnt;
        private int spiedMarshallerParseCallCnt;

        int getSpiedMarshallerStreamCallCnt() {
            return spiedMarshallerStreamCallCnt;
        }

        int getSpiedMarshallerParseCallCnt() {
            return spiedMarshallerParseCallCnt;
        }

        @Override
        public <I, O> ClientCall<I, O> interceptCall(MethodDescriptor<I, O> method, CallOptions callOptions,
                                                     Channel next) {
            final MethodDescriptor<I, O> methodDescriptor = method.toBuilder().setRequestMarshaller(
                    new PrototypeMarshaller<I>() {
                        @Nullable
                        @Override
                        public I getMessagePrototype() {
                            return null;
                        }

                        @Override
                        public Class<I> getMessageClass() {
                            return null;
                        }

                        @Override
                        public InputStream stream(I value) {
                            spiedMarshallerStreamCallCnt++;
                            return method.getRequestMarshaller().stream(value);
                        }

                        @Override
                        public I parse(InputStream inputStream) {
                            return null;
                        }
                    }).setResponseMarshaller(
                    new PrototypeMarshaller<O>() {
                        @Nullable
                        @Override
                        public O getMessagePrototype() {
                            // TODO: to return proper Message
                            return null;
                        }

                        @Override
                        public Class<O> getMessageClass() {
                            return null;
                        }

                        @Override
                        public InputStream stream(O o) {
                            return null;
                        }

                        @Override
                        public O parse(InputStream inputStream) {
                            spiedMarshallerParseCallCnt++;
                            return method.parseResponse(inputStream);
                        }
                    }).build();
            return next.newCall(methodDescriptor, callOptions);
        }
    }
```
related: https://github.com/line/armeria/blob/5543dc1083d870a88585dda895ae4efc7d54a83f/grpc/src/main/java/com/linecorp/armeria/internal/common/grpc/GrpcMessageMarshaller.java#L245-L247

Reply:
The request type `unaryCall` is `SimpleRequest` which we can use:
```java
new PrototypeMarshaller<I>() {
    @Nullable
    @Override
    public I getMessagePrototype() {
        return (I) SimpleRequest.getDefaultInstance();
    }

    @Override
    public Class<I> getMessageClass() {
        return (Class<I>) SimpleRequest.class;
    }
    ...
```

Reply:
Oh, It's simple.
Thank you. It works.

Reply:
Applied feedbacks.
Thanks.